Perf optimization: Make creation of uninstall actions run in parallel

erdembayar opened the pull request:
## Bug

Fixes: [Issue#9956](https://github.com/NuGet/Home/issues/9956)
Regression: No  

## Fix

Details: When we uninstall package from solution there 2 stages. First `create uninstallActions`, then `execute uninstallActions` . 
We already optimized `execute uninstallActions` part so it's at least 2 times faster for large solutions like `OrchardCore`, but I missed optimizing `create uninstallActions` part and I just implemented it.

### Result
With this optimization has clear benefit for .Netcore projects. OrchardCore solution get **additional** boost, now it's another **2 times faster**, almost **50% reduction in runtime**.  Just compare with all of current improvement efforts it was [2 min 1 sec](https://gist.github.com/erdembayar/2ba554f169e4f078baa61976affa62d4)  2 months ago, but now it's **30-35 second range consistently 😄** , **total 4 times faster now**.

---------------------------------------------------------------------------------
Below are **new** measurements after adding improvement for evaluate dependency graph only once. 
If you compare with previous improvement measurements I can see extra few seconds improvements now.
**OrchardCore (158 projects) 
Before optimization:**
[1 min 10 sec](https://gist.github.com/erdembayar/6d575837cc44d5eacc07387d823638fd)
[1 min 11 sec](https://gist.github.com/erdembayar/fe7489ab08f058d9bf8b43b54d9af015)
[1 min 14 sec](https://gist.github.com/erdembayar/3c756f0c6aac4e80e8c15eee26eeddc9)
**After optimization:**
[30 sec](https://gist.github.com/erdembayar/cdf0c94f863afcb25dc75f812753a746)
[35 sec](https://gist.github.com/erdembayar/83b99ffdf9746c2116896699c4bf36aa)
[34 sec](https://gist.github.com/erdembayar/55ee7d377bcf9431045174c4b0782703)

**SmallNetCoreProject(4 projects)
Before optimization:**
[1.32 sec ](https://gist.github.com/erdembayar/c29a0f543a4bdd8da0637b0f2c35b39e)
[1.01 sec](https://gist.github.com/erdembayar/e97f50de18733c23515a028bb226d874)
**After optimization:**
[0.35 sec](https://gist.github.com/erdembayar/57f6c50449de1b562f4cf909c5b65964)
[0.72 sec](https://gist.github.com/erdembayar/0be0a61a38fd72a64a8f8951f65151e4)

---------------------------------------------------------------------------------
Below are **old** measurements before latest improvement for evaluate dependency graph only once.

OrchardCore (158 projects) 
Before optimization:
[1 min 5 sec](https://gist.github.com/erdembayar/be781bb927e915f1f299701c28581b3c)
[1 min 3 sec](https://gist.github.com/erdembayar/45fb08536a4cef1c26d5f36018d597d2)
[1 min 1 sec](https://gist.github.com/erdembayar/c5bc1071058d742411dce2ad7fef5919)
After optimization:
[41 sec](https://gist.github.com/erdembayar/3015428a682f773c08ab7691bb13046e)
[36 sec](https://gist.github.com/erdembayar/a460255e456f1bc9099810bac9c0afee)
[37 sec](https://gist.github.com/erdembayar/958c5ec8c7c7af085014ae5686cfb95a)
[39 sec](https://gist.github.com/erdembayar/efcfd285dc35832adbbe33e00a7eb9d7)

SmallNetCoreProject(4 projects)
Before optimization:
[1.45 sec](https://gist.github.com/erdembayar/2d5e59ac281dc5a2f3c58f95d06442ba)
[1.34 sec](https://gist.github.com/erdembayar/11cbc65d5a68db180c89f89eb4e31098)

After optimization:
[0.80 sec](https://gist.github.com/erdembayar/5d3f3311d404c56194742e16bb733a2a)
[1.13 sec](https://gist.github.com/erdembayar/5a9559bec12505a363a59fb22efe237a)
[0.95 sec](https://gist.github.com/erdembayar/a2762ea8bb2ede678861ebc91ceb4851)

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:  Manual testing several times for Large/Medium/small projects. Added both positive and negative unit tests.

erdembayar committed:
Made creation of uninstall actions run in parallel.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/INuGetProjectManagerService.cs:
```diff
@@ -46,7 +46,7 @@ public interface INuGetProjectManagerService : IDisposable
 
         ValueTask<IReadOnlyList<ProjectAction>> GetUninstallActionsAsync(
-            string projectId,
+            IReadOnlyCollection<string> projectIds,
```
This is an API break that'd break Codespaces support. 

cc @dtivel @rrelyea for review here. 

We probably need a PublicAPI file here too to help us manage this. 
cc @zivkan

zivkan replied:
https://github.com/NuGet/NuGet.Client/pull/3708

erdembayar replied:
@nkolev92  I created threat on Codespace channel and tagged them. 
If it's the case then I can remedy this by create new separate api signature here.

erdembayar replied:
I talked with @rrelyea and [this break is ok](https://teams.microsoft.com/l/message/19:f5bb1554b1004a13897f486520d561a8@thread.skype/1602611417102?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=ad34ac55-6785-4e7c-82c3-7ba405d30c3c&parentMessageId=1602028186751&teamName=NuGet&channelName=NuGet%20for%20VS.Oe&createdTime=1602611417102).

nkolev92 commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+            return uninstallActionsForProjects;
+        }
+
+        private async Task GetProjectNuGetUninstallProjectActionsAsync(ProjectUninstallActionSpec projectUninstallActionSpec)
```
I'm not sure this is parallelizing the restore runs as we want them to be. 

This calls 

```cs
        private async Task<IEnumerable<NuGetProjectAction>> PreviewUninstallPackageInternalAsync(NuGetProject nuGetProject, Packaging.PackageReference packageReference,
```

which calls:

```cs
        public async Task<BuildIntegratedProjectAction> PreviewBuildIntegratedProjectActionsAsync(
            BuildIntegratedNuGetProject buildIntegratedProject,
            IEnumerable<NuGetProjectAction> nuGetProjectActions,
            INuGetProjectContext nuGetProjectContext,
            CancellationToken token)
```
which runs restore.

erdembayar replied:
@nkolev92 
I believe it's ok, because we're already doing it for `install action`.

Here `PreviewBuildIntegratedProjectActionsAsync`   [Line 2651](https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697) is just simple façade for https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697-L2702

Here `PreviewBuildIntegratedProjectActionsAsync` on [Line 2697](https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697) is already called for `install action` from https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1601

Here below is how call stack looks like until  we reach above [NuGetPackageManager.cs#L1601](https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1601) for install action: 
1. If it's 'install action'?
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L776

2. Then call this `GetInstallActionsAsync` method:
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L780

3. `GetInstallActionsAsync` method is here:
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs#L256

4. From inside here call method `PreviewProjectsInstallPackageAsync`:
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs#L294

5.  We call method `PreviewBuildIntegratedProjectsActionsAsync` : https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1598-L1601

nkolev92 replied:
In 4, preview Project*s*, plural, is calling Project, singular. 

In the current implementation you are calling restore individually. If I'm following the code correctly, uninstall will end up calling: https://github.com/NuGet/NuGet.Client/blob/087a7113ad70ae4d4c538eba0f387271af0c649b/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L141-L151 and not 
https://github.com/NuGet/NuGet.Client/blob/087a7113ad70ae4d4c538eba0f387271af0c649b/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L194-L204.

erdembayar replied:
@nkolev92 
Uninstall action is calling correct correct dgSpec method (plural one).
Here call stack starting parallel job call:
1. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2057

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2070-L2075

2. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2169

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2186-L2188

3. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2651

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2678-L2688

4. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2858-L2870

5. Here correct plural dgSpec method.
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L194

nkolev92 replied:
That's because:

PreviewBuildIntegratedProjectActionsAsync calls PreviewBuildIntegratedProjectsActionsAsync

Basically from: 
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2169

to

https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2651

to 

https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697

If you put a break point in DependencyGraphRestoreUtility.cs.PreviewRestoreProjectsAsync, you will see that it's called twice, while it should be called once.

erdembayar replied:
@nkolev92  
Ok, done. 
Please check. I see another few seconds gain on large project like OrchardCore. 😄

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs:
```diff
@@ -336,44 +337,47 @@ public async ValueTask ExecuteActionsAsync(IReadOnlyList<ProjectAction> actions,
             Assumes.NotNull(packageIdentity);
             Assumes.False(packageIdentity.HasVersion);
             Assumes.NotNull(_state.SourceCacheContext);
             Assumes.NotNull(_state.ResolvedActions);
```
Remove this statement.  `ResolvedActions` is never `null`, and you added the correct replacement below.

erdembayar replied:
Ok. Resolved.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs:
```diff
@@ -336,44 +337,47 @@ public async ValueTask ExecuteActionsAsync(IReadOnlyList<ProjectAction> actions,
 
-            INuGetProjectContext projectContext = await ServiceLocator.GetInstanceAsync<INuGetProjectContext>();
-            (bool success, NuGetProject? project) = await TryGetNuGetProjectMatchingProjectIdAsync(projectId);
+            _state.PackageIdentity = packageIdentity;
```
Remove.  This is for `IsDirectInstall(...)` for installing a package.

erdembayar replied:
Ok. Removed.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+        /// <param name="nuGetProjectContext"></param>
+        /// <param name="token"></param>
+        /// <returns></returns>
+        public async Task<IEnumerable<(NuGetProject, NuGetProjectAction)>> PreviewProjectsUninstallPackageAsync(
```
Simplify the return type to `Task<IEnumerable<NuGetProjectAction>>`.  [`NuGetProjectAction.Project`](https://github.com/NuGet/NuGet.Client/blob/65abcc61f73ab3d87928495f789ec74011edde7f/src/NuGet.Core/NuGet.PackageManagement/NuGetProjectAction.cs#L53) already indicates the target project.

erdembayar replied:
Ok. Done.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+                                                nuGetProjectContext,
+                                                uninstallActionsForProjects,
+                                                token)
+                                            );
```
Indent 1 level, which is 4 spaces, from the previous line (2041).  Also move `);` to end of previous line.

erdembayar replied:
Ok. Moved.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+            };
+            var actionBlock = new ActionBlock<ProjectUninstallActionSpec>(GetProjectNuGetUninstallProjectActionsAsync, options);
+
+            foreach (var nuGetProject in nuGetProjects)
```
Use explicit type.

erdembayar replied:
Ok. Fixed.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+                projectUninstallActionSpec.NuGetProjectContext,
+                projectUninstallActionSpec.Token);
+
+            foreach (var projectUninstallAction in projectUninstallActions)
```
Use explicit type.

erdembayar replied:
Ok. Fixed.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+                throw new ArgumentNullException(nameof(nuGetProjectContext));
+            }
+
+            var uninstallActionsForProjects = new ConcurrentBag<(NuGetProject, NuGetProjectAction)>();
```
```C#
var uninstallActionsForProjects = new ConcurrentBag<NuGetProjectAction>();
```

erdembayar replied:
Fixed.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+        /// <param name="nuGetProjectContext"></param>
+        /// <param name="token"></param>
+        /// <returns></returns>
+        public async Task<IEnumerable<(NuGetProject, NuGetProjectAction)>> PreviewProjectsUninstallPackageAsync(
```
This is a new public API.  Can you add some unit tests, both positive and negative?

erdembayar replied:
Ok. I added unit tests for both happy path and unhappy path.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs:
```diff
@@ -501,7 +495,7 @@ private static string CreateProjectActionId()
 
             Dictionary<string, NuGetProject>? projects = (await solutionManager.GetNuGetProjectsAsync())
-                .ToDictionary(project => project.GetMetadata<string>(NuGetProjectMetadataKeys.ProjectId), _ => _);
+                .ToDictionary(project => project.GetMetadata<string>(NuGetProjectMetadataKeys.ProjectId), _ => _, PathUtility.GetStringComparerBasedOnOS());
```
This doesn't look right.  Project ID's are GUID's not file system paths.  I think this should be `StringComparer.OrdinalIgnoreCase` like [it was](https://github.com/NuGet/NuGet.Client/pull/3703/files#diff-b1db843d684b7d235cd2b07f4917eafaL556).

That said, I'm surprised you'd need it.  Do you have multiple projects with the same project ID but differing in casing?

erdembayar replied:
Sorry, I didn't realize it was just GUID. I'll fix it.

erdembayar replied:
Fixed it.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+                    projectUninstallActionSpec.PackageId, projectUninstallActionSpec.NuGetProject.GetMetadata<string>(NuGetProjectMetadataKeys.Name)));
+            }
+
+            var projectUninstallActions = await PreviewUninstallPackageInternalAsync(
```
Use explicit type.

erdembayar replied:
Addressed it.

erdembayar committed:
Address code review comments by @dtivel

erdembayar committed:
Add unit tests for covering new `PreviewProjectsUninstallPackageAsync` method in NuGetPackageManager.

erdembayar committed:
Address 1 more code review comment from dtivel.

erdembayar committed:
Add unhappy path unit test.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs:
```diff
@@ -423,17 +421,7 @@ public async ValueTask ExecuteActionsAsync(IReadOnlyList<ProjectAction> actions,
-
-                projects.Add(project);
-            }
+            var projects = await GetProjectsAsync(projectIds, cancellationToken);
```
Use explicit type.

erdembayar replied:
Ok. Addressed.

dtivel commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+
+            // Step-1: Get the packageIdentity corresponding to packageId and check if it exists to be uninstalled
+            var installedPackages = await projectUninstallActionSpec.NuGetProject.GetInstalledPackagesAsync(projectUninstallActionSpec.Token);
+            var packageReference = installedPackages.FirstOrDefault(pr => pr.PackageIdentity.Id.Equals(projectUninstallActionSpec.PackageId, StringComparison.OrdinalIgnoreCase));
```
Use explicit types instead of `var`.

erdembayar replied:
Ok. Addressed.

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs:
```diff
@@ -2316,6 +2316,694 @@ public async Task GetInstalledPackages_WithNominationUpdate_ReloadsCache()
+                // Project
+                int numberOfProjects = 4;
+                var projectCache = new ProjectSystemCache();
+                IVsProjectAdapter projectAdapter = (new Mock<IVsProjectAdapter>()).Object;
```
`Mock.Of<IVsProjectAdapter>()` is more concise.

erdembayar replied:
Ok. Addressed, I copied previous pattern from another existing unit test without much thinking. This one makes more sense.

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs:
```diff
@@ -2316,6 +2316,694 @@ public async Task GetInstalledPackages_WithNominationUpdate_ReloadsCache()
+
+                var testNuGetProjectContext = new TestNuGetProjectContext() { EnableLogging = true };
+                var netCorePackageReferenceProjects = new List<NetCorePackageReferenceProject>();
+                var prevProj = string.Empty;
```
`string prevProjectFullPath = null;`

erdembayar replied:
Ok. Addressed.

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs:
```diff
@@ -2316,6 +2316,694 @@ public async Task GetInstalledPackages_WithNominationUpdate_ReloadsCache()
+                var sources = new List<PackageSource>();
+                var packageA_Version100 = new SimpleTestPackageContext("packageA", "1.0.0");
+                var packageB_Version100 = new SimpleTestPackageContext("packageB", "1.0.0");
+                var packageA = packageA_Version100.Identity;
```
Use `var` when the initialization type is trivially obvious (i.e.:  a literal value, a `new` expression`, a cast); otherwise, use the explicit type.  This is rule 10 from the [coding guidelines](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md#c-coding-style).

Use `var`:
*  `var a = 1; // int`
*  `var b = false; // bool`
*  `var c = new object(); // object`
*  `var d = (IDisposable)e; // IDisposable`
*  `var f = g as H; // H`

It makes the code more quickly grokkable.  Also, mentally inferring the actual type from `var` can be challenging in a browser-based code review.  I may have specific feedback if I know the actual types in use.

This feedback applies to all changes in this file.

erdembayar replied:
Ok. Thank you for your feedback.
Funny problem is 😆  back in sometime ago(not at MS) I used to receive exact opposite feedback, I had to replace explicit types with `var` to pass code review. Now it's opposite, old habits dies hard. No problem, I'll fix it. 😄

erdembayar replied:
How about is it ok to not use `var` at all?

dtivel replied:
I understand.  Me too.  Even within our team there are some who prefer "var always" and some who don't.  The point of the rule is to achieve consistency.

You shouldn't avoid `var` altogether.  The rule is to use `var` but only when the right-hand side is "obvious" (explained above).  The intent is to be terse but not at the expense of readability.

erdembayar replied:
Addressed.

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs:
```diff
@@ -2316,6 +2316,694 @@ public async Task GetInstalledPackages_WithNominationUpdate_ReloadsCache()
+                    netCorePackageReferenceProjects,
+                    actions,
+                    testNuGetProjectContext,
+                    new SourceCacheContext(),
```
`SourceCacheContext` is disposable.  Wrap usage with a `using` statement.

This feedback applies globally (e.g.:  other tests).

erdembayar replied:
Addressed.

erdembayar committed:
Address more code review comments by dtivel.

erdembayar committed:
Fix rebasing to dev.

erdembayar committed:
Merge branch 'dev' into dev-eryondon-parallelize-createactions-foruninstall

erdembayar committed:
Add another perf improvement for BuildIntegratedNuGetProject type projects.
BuildIntegratedNuGetProject project can batch eval of DependencyGraphRestoreUtility.PreviewRestoreProjectsAsync.

erdembayar committed:
Fix logic error for non-builtIntegrated projects.

erdembayar commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2019,7 +2028,91 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+                results.AddRange(uninstallActions);
+            }
+
+            foreach (var target in otherTargetProjectsToUpdate)
```
@nkolev92 Is it ok to run this parallel for non-builtIntegrated projects. Actually we don't do it for install action for this types..

nkolev92 replied:
We don't do it for non-buildintegrated projects, so we shouldn't do it for uninstall either.

erdembayar replied:
Ok. Removed.

erdembayar replied:
Removed.

nkolev92 commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2019,7 +2028,91 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+
+            var results = new List<NuGetProjectAction>();
+
+            // BuildIntegratedNuGetProject type projects are now supports parallel preview action for faster performance.
```
I don't think provides much value. It's pretty obvious that it affects multiple projects, hence the name.

erdembayar replied:
Removed.

nkolev92 commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2019,7 +2028,91 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+            var buildIntegratedProjectsToUpdate = new List<BuildIntegratedNuGetProject>();
+            var otherTargetProjectsToUpdate = new List<NuGetProject>();
+
+            foreach (var proj in nuGetProjects)
```
You can make this one line with a where statement + ToList.

erdembayar replied:
I know. But If I remember correctly previously I received code review stating this one is more performant than that on during another perf [PR#3559](https://github.com/NuGet/NuGet.Client/pull/3559) review. I can't find that exact comment since there're 211 comments from different people.

nkolev92 reviewed (commented):
looks better. 
We shouldn't parallelize non-build integrated projects.

erdembayar committed:
Remove optimization for msbuild package.config type projects since they're not part of this optimization scope.

erdembayar committed:
Remove unnecessary comment.

erdembayar committed:
Merge branch 'dev' into dev-eryondon-parallelize-createactions-foruninstall

erdembayar committed:
Merge branch 'dev' into dev-eryondon-parallelize-createactions-foruninstall

erdembayar committed:
Fix formatting.

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs:
```diff
@@ -2316,6 +2316,703 @@ public async Task GetInstalledPackages_WithNominationUpdate_ReloadsCache()
+                Assert.True(builtIntegratedActions.All(b => b.RestoreResult.Success));
+                Assert.True(builtIntegratedActions.All(b => !b.RestoreResult.LogMessages.Any())); // There should be no error or warning.
+                List<string> uninstalledLogs = testNuGetProjectContext.Logs.Value.Where(l => l.StartsWith("Successfully uninstalled ")).ToList();
+                Assert.Equal(uninstalledLogs.Count(), 1); // There is bug in Package manager log it only mention 1 project is "Successfully uninstalled " instead of 4.
```
Is there a tracking issue for this bug?
Is it an actual bug or is it inconvenient?

erdembayar replied:
Sorry, it's more of "inconvenient" part. I just called it bug. I'll change the wording.

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs:
```diff
@@ -2316,6 +2316,703 @@ public async Task GetInstalledPackages_WithNominationUpdate_ReloadsCache()
         }
 
+        [Fact]
+        public async Task TestPackageManager_PreviewProjectsUninstallPackageAsync_AllProjects_Success()
```
Consider the What_UnderWhichConditions_ExpectedREsult pattern. 

Now you are slightly missing the under which conditions part. +

erdembayar replied:
Ok. Addressed.

nkolev92 reviewed (approved):
Product code looks good. 

Couple of little comments on the tests.

erdembayar committed:
Address review by Nikolche.

erdembayar committed:
Merge branch 'dev' into dev-eryondon-parallelize-createactions-foruninstall

erdembayar committed:
Fix flaky test due.

erdembayar committed:
Remove unused variables from test.
